openshift-201/rh201-pod-auto-scale.md

```diff
@@ -0,0 +1,283 @@
+
+Give it a few minutes and check the vpa status sections for recommendations.
+
+
+```yaml
+oc get vpa vpa-recommender -o yaml
+
+ status:
+    conditions:
+    - lastTransitionTime: "2022-04-11T17:35:54Z"
+      status: "True"
+      type: RecommendationProvided
+    recommendation:
+      containerRecommendations:
+      - containerName: hello-world-nginx
+        lowerBound:
+          cpu: 25m
+          memory: 262144k
+        target:
+          cpu: 35m
+          memory: 262144k
+        uncappedTarget:
+          cpu: 35m
+          memory: 262144k
+        upperBound:
+          cpu: 66m
+          memory: 262144k
+```
+
+
+The output shows the recommended resources under target.
+The minimum recommended resources under lowerBound. 
+The highest recommended resources under upperBound.
+The most recent resource recommendations under uncappedTarget.
+
+With the recommendations, you can edit the workload object to add CPU and memory requests, then delete and redeploy the pods using the recommended resources.
+
+If the VPA updateMode uses something other than `off` then the lowerBound and upperBound values are used to determine if a pod needs to be updated. If a pod has resource requests below the lowerBound values or above the upperBound values, the VPA terminates and recreates the pod with the target values.
+
+### Update Mode Auto
+
+Lets create an VPA in auto mode and send traffic to the hello-world-nginx pods and observe what happens.
+
+VPAs in Auto updateMode won't work with HPA using the same CPU and memory metrics because it would cause a race condition. If HPA objects still exists from previous labs lets delete them to focus on VPAs.
+
+Then lets scale our hello-world-nginx deployment replicas up to three so the VPA can re-deploy our pods if needed.
+
+Then lets create another VPA in Auto updateMode:
+
+```yaml
+cat <<EOF | oc apply -f -
+apiVersion: autoscaling.k8s.io/v1
+kind: VerticalPodAutoscaler
+metadata:
+  name: vpa-auto
+spec:
+  targetRef:
+    apiVersion: "apps/v1"
+    kind:       Deployment 
+    name:       hello-world-nginx
+  updatePolicy:
+    updateMode: "Auto"
+EOF
+```
+Again update the load-test deployment environment variable REQUESTS to a different number to trigger a redeployment and the load-test pod to start sending traffic again.
+
+Give it a few minutes and observe the VPA and the hello-world-nginx pods. We should seem them re-deploy based on updated values from the VPA.
+
+### Summary 
+
+Have a think what would be good for app in a production environment. Maybe just getting recommendations to review might be good then you update manually. Having Auto mode on will terminate and restart pods and containers which is maybe something your application can't handle in a production environment.
+
+VPAs are not aware of OpenShift cluster infrastructure variables such as node size in terms of memory and CPU. Therefore, it doesn't know whether a recommended pod size will fit your node. 
```

Comment:
Maybe also note it doesn't know about project quotas?

Reply:
Good call! I added that in.